Update docker image to Bioconductor 3.19, make it multiplatformable

jashapiro opened the pull request:
Initiated as part of https://github.com/AlexsLemonade/rstudio-server/issues/10
Closes #748

Here I updated the renv file to use bioconductor 3.18 and R 4.3.2, updating a number of packages along the way.

I then updated the Dockerfile, changing the base to a Bioconductor-based docker image. This allows us to have binary installs of bioconductor packages when building `amd64` images, which is a nice little time-saver.

In addition, I took the opportunity to make the image multiplatform compatible. This makes it run much more efficiently on Apple/arm processors, which is handy! To do this, we have to compile a few things that we previously downloaded as binaries, but this seemed overall worth the cost. I minimized that cost by adding a multi-stage build process where we build platform-specific binaries like salmon in a separate image, then only copy over the required files. By doing this copy at the end of the Docker setup, we run the compilation and R package installation in parallel, which speeds up build time.

Building the arm64 binary is still too slow to do with GHA, in part because the bioconductor packages are not available as binaries, so I am only doing the amd64 build still. We might want to consider building manually for releases so we can have 

I did experiment with Docker build cloud, but I could not even get that to fully work locally: While I could build a single binary, trying to multiplatform builds gave the following error: `ERROR: cloud pull for multi-node builds currently not supported`. Maybe sometime later, but the R installation is slow enough that we would eat up build minutes pretty fast! It also seems like the GHA integration is still a bit experimental: the instructions are not nearly as detailed as I would have hoped.

jashapiro commented:
I want to note that I have not checked all of the modules to see that things still work as expected after the updates here. If we are happy with this setup, I will file a separate issue to check that and rebuild everything before we next use the modules!

jaclyn-taroni reviewed (commented):
I'm fine with this strategy. I would like @davidsmejia to take the last look at it. I'd also like to see at least this section of the contributing guidelines https://github.com/AlexsLemonade/training-modules/blob/master/CONTRIBUTING.md#pushing-to-dockerhub-via-github-actions get updated in this PR.

jashapiro commented:
> I'm fine with this strategy. I would like @davidsmejia to take the last look at it. I'd also like to see at least this section of the contributing guidelines https://github.com/AlexsLemonade/training-modules/blob/master/CONTRIBUTING.md#pushing-to-dockerhub-via-github-actions get updated in this PR.

I had only updated the testing build action here, not the build and push action, but I can make those updates in this PR as well to keep things in sync.

As you might have noticed in the build action, I was thinking I might change the naming scheme to `ccdl/training_rstudio` rather than `training_dev`, and use that for the training-tagged releases as well, but that requires a few docs changes as well (I will file the appropriate issues).
We can keep the automatic build and push for `latest`, but I was thinking I would also add tagged builds for releases to at least partially replace [this release task](https://github.com/AlexsLemonade/training-admin/blob/master/workshop-task-templates/docker-tag-and-push.md), like we do with `scpcaTools`.

If we want to keep multiplatform builds, we would still have to do those manually, in the same way we had been doing the current builds for releases.

jaclyn-taroni commented:
> I had only updated the testing build action here, not the build and push action, but I can make those updates in this PR as well to keep things in sync.

Yes, we should be keeping things in sync. If you're just building, I'm confused about why you're loading secrets?

jashapiro commented:
> > I had only updated the testing build action here, not the build and push action, but I can make those updates in this PR as well to keep things in sync.
> 
> Yes, we should be keeping things in sync. If you're just building, I'm confused about why you're loading secrets? **Edit:** Is accessing the cache still relevant if one is not using Docker build cloud?

The secrets/login are so we can use dockerhub for the cache. The hope is that using this rather that the GHA cache would save time overall, as the GHA cache is relatively small, and this image is... not.

jashapiro commented:
I have now updated the image to R 4.4 and Bioconductor 3.19. I also consolidated the two GHA workflows into one, with automatic tagging and pushing of images when we tag a version of the repo for training. Development versions are now tagged with `edge`, while the most recent training tag gets the `latest` tag. 

I updated the CONTRIBUTING doc to reflect these changes as well. If this looks good, I think it may make sense to merge this version in and make any changes related to final processing in a separate PR.

jaclyn-taroni reviewed (approved):
LGTM
